### ISSUE TITLE
Auto domain based on a supplied extent

### ISSUE THREAD

**Issue by chrisprice** [d3fc/d3fc#1329]
Could we allow specifying an `xExtent`/`yExtent` on Cartesian to be used to auto calculate a domain?

**Comment by chrisprice** [comment]
A bit more background on the motivation. I'm trying to make the getting started experience easier for new users without compromising on our pluggable component model. The most [simple cartesian chart example](https://github.com/d3fc/d3fc/blob/master/packages/d3fc-chart/examples/chart.js#L1-L41) we have is 40 lines long of which 20% of it relates to setting up `extent`/`domain` both of which (I think!) must sound like alien concepts to a new user.

Additionally, there is only 1 line which contributes novel configuration i.e. setting the `pad` on the `xExtent`, the rest is just plumbing (creating `extent` and passing it to `domain`) or duplicating existing values (`accessors`).

If we want to eliminate these lines, we need to answer -

1. How do we retrieve the set of `accessors` from the series?
2. How do we choose between `extentLinear` and `extentTime`?
3. Where does the *magic* live?
4. How do we stop it feeling like magic?

**Comment by chrisprice** [comment]
Possible solutions/thoughts -

1. Extend the series "interface" to additionally include `accessors` which would in most cases return the tuple of `mainValue`/`crossValue`. `multiSeries` would aggregate the set of series `accessors`. `gridlines` would end up returning an empty array. Downsides of this approach would be polluting the series "interface" and probable duplication of accessor functions, marginally increasing the runtime of the extent calculation.

2. I'd rather we didn't allow the `extent` implementation to be pluggable. I think explicitly setting a `domain` is a clearer alternative (or omitting/setting to `null` to trigger extent behaviour). That allows us to rebind the `extent` implementation methods e.g. `pad`.

**Comment by ColinEberhardt** [comment]
I'm broadly happy with the suggestions for (1) and (2). 

> I'd rather we didn't allow the extent implementation to be pluggable.

The way it works at the moment, it isn't really 'plugged' into the cartesian chart.

Is it worth looking at simplifying the standard update pattern for the chart? again the extent interface causes some confusion - each update you re-render the chart via a standard D3 data join. However, you also have to explicitly update the domain to reflect the change in extent. This is super confusing!

**Comment by chrisprice** [comment]
Cool and good point on simplifying update. So the change would be if `xDomain`/`yDomain` is not set to a non-null value, `cartesianChart` will calculate their values using an XY pair of internal `extentLinear`/`extentTime` implementations. 

I think that's a suitable answer to 3/4 but raises 2 further questions -

5. What is the read behaviour of `xDomain`/`yDomain` values i.e. should we return the extent calculated values? This would be the only way a user could access these values without doing something obscure like asking for `ticks`. However, this could be confusing as the calculated values will only be available once the chart has been rendered and the value you write to the property is not what you'll read back.

6. Manually choosing between `extentLinear`/`extentTime` will require an additional property (e.g. `xExtentType` = `linear` | `time`). This will add an extra configuration line unless we default the selection based upon the passed scale which is non-trivial to derive. An alternative would be to further extend the constructor properties to include `xExtent`/`yExtent`.

**Comment by chrisprice** [comment]
And some suggestions -

5. I think I'd rather avoid the confusing behaviour of sometimes returning the calculated values. If you want them, you can always just use extent yourself and then set the domain explicitly.

6. I'm tempted to invest some time refactoring `d3fc-extent` to create a new `extent` component which always applies `valueOf` to the values. I think in the 99% case that will give the right result. For the 1% we can still have `extentLinear` & `extentTime`, the user can then explicitly use one or the other and then explicitly set the domain as above.

**Comment by ColinEberhardt** [comment]
I agree with (5), looking at the various examples I've built with D3FC there are very few occasions where I've needed the computed extent.

Re (6) good idea, what would be the 1% case where this might fail?

**Comment by chrisprice** [comment]
I was thinking of an edge case where `valueOf` was overridden in a weird way for a custom object e.g. it always returned the same value. I guess that only applies to a category scale.

**Comment by ColinEberhardt** [comment]
>  I guess that only applies to a category scale.

In which case you don't need to use extent :-)
